tests/test_transform_hoist_variables.py

```diff
@@ -0,0 +1,276 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+"""
+A selection of tests for the (generic) hoist variables functionalities.
+"""
+from pathlib import Path
+import pytest
+import numpy as np
+
+from conftest import available_frontends, jit_compile, clean_test
+from loki import FindNodes, Scheduler
+from loki import ir, is_iterable
+from loki.transform import (HoistVariablesAnalysis, HoistVariablesTransformation,
+                            HoistTemporaryArraysAnalysis, HoistTemporaryArraysTransformationAllocatable)
+
+
+@pytest.fixture(scope='module', name='here')
+def fixture_here():
+    return Path(__file__).parent
+
+
+@pytest.fixture(name='config')
+def fixture_config():
+    """
+    Default configuration dict with basic options.
+    """
+    return {
+        'default': {
+            'mode': 'idem',
+            'role': 'kernel',
+            'expand': True,
+            'strict': True,
+        },
+        'routine': [
+            {
+                'name': 'driver',
+                'role': 'driver',
+                'expand': True,
+            },
+            {
+                'name': 'another_driver',
+                'role': 'driver',
+                'expand': True,
+            },
+        ]
+    }
+
+
+def compile_and_test(scheduler, here, a=(5,), test_name=""):
+    """
+    Compile the source code and call the driver function in order to test the results for correctness.
+    """
+    assert is_iterable(a) and all(isinstance(_a, int) for _a in a)
+    for item in scheduler.items:
+        if "driver" in item.name:
+            suffix = '.F90'
+            if test_name != "":
+                item.source.path = Path(f"{item.source.path.stem}_{test_name}")
+            module = jit_compile(item.source, filepath=item.source.path.with_suffix(suffix).name, objname=None)
+            for _a in a:
+                parameter_length = 3
+                b = np.zeros((_a,), dtype=np.int32, order='F')
+                c = np.zeros((_a, parameter_length), dtype=np.int32, order='F')
+                module.Transformation_Module_Hoist.driver(_a, b, c)
+                assert (b == 42).all()
+                assert (c == 11).all()
+            clean_test(filepath=here.parent / item.source.path.with_suffix(suffix).name)
+
+
+def check_arguments(scheduler, subroutine_arguments, call_arguments):
+    """
+    Check the subroutine and call arguments of each subroutine.
+    """
+    for item in scheduler.items:
```

Comment:
Instead of loop+if/elif you could just access the items directly. Something like `scheduler.item_map['#driver'` or scheduler.item_map['#device2'].